File: day022_counting_sort/primary_counting_sort/solution.py
from typing import List

class Solution:
    # Optimized: O(N + K) Time, O(N + K) Space
    def countingSort(self, nums: List[int]) -> List[int]:
        if not nums:
            return nums
        
        # 1. Find the range (handling negative numbers with an offset)
        min_val = min(nums)
        max_val = max(nums)
        range_of_elements = max_val - min_val + 1
        
        # 2. Initialize the count array
        count = [0] * range_of_elements
        output = [0] * len(nums)
        
        # 3. Store the frequency of each element
        for num in nums:
            count[num - min_val] += 1
            
        # 4. Modify count[i] to contain the actual position in output
        for i in range(1, len(count)):
            count[i] += count[i - 1]
            
        # 5. Build the output array (traverse backwards for stability)
        for i in range(len(nums) - 1, -1, -1):
            output[count[nums[i] - min_val] - 1] = nums[i]
            count[nums[i] - min_val] -= 1
            
        return output

File: day022_counting_sort/primary_counting_sort/solution.java
import java.util.Arrays;

class Solution {
    // O(N + K) Time, O(N + K) Space
    public int[] countingSort(int[] nums) {
        if (nums == null || nums.length == 0) return nums;

        int minVal = nums[0];
        int maxVal = nums[0];
        for (int num : nums) {
            if (num < minVal) minVal = num;
            if (num > maxVal) maxVal = num;
        }

        int range = maxVal - minVal + 1;
        int[] count = new int[range];
        int[] output = new int[nums.length];

        // Frequency Count
        for (int num : nums) {
            count[num - minVal]++;
        }

        // Cumulative Count for Stability
        for (int i = 1; i < range; i++) {
            count[i] += count[i - 1];
        }

        // Output Construction
        for (int i = nums.length - 1; i >= 0; i--) {
            output[count[nums[i] - minVal] - 1] = nums[i];
            count[nums[i] - minVal]--;
        }

        return output;
    }
}